src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php

```diff
@@ -199,9 +200,18 @@ private function willMakeFinalResponseUncacheable(Response $response): bool
      * If the value is lower than the currently stored value, we update the value, to keep a rolling
      * minimal value of each instruction. If the value is NULL, the directive will not be set on the final response.
      */
-    private function storeRelativeAgeDirective(string $directive, ?int $value, int $age)
+    private function storeRelativeAgeDirective(string $directive, ?int $value, int $age, bool $isHeuristicallyCacheable)
     {
         if (null === $value) {
+            if ($isHeuristicallyCacheable) {
+                /*
+                 * See https://datatracker.ietf.org/doc/html/rfc7234#section-4.2.2
+                 * This particular response does not require maximum lifetime; heuristics might be applied.
+                 * Other responses, however, might have more stringent requirements on maximum lifetime.
+                 * So, return early here so that the final response can have the more limiting value set.
+                 */
+                return;
+            }
             $this->ageDirectives[$directive] = false;
```

Comment:
i wonder if the code below that updates the directive should not overwrite `false` with the $value? if one fragment does not specify max-age, but others do, don't we want the lowest value of max-age output, rather than no max-age at all?

btw, just learned that `min(false, 3)` is valid and gives boolean false: https://3v4l.org/Ztr36 - color me surprised... but that does not matter as we have the `if false!==` condition.

Reply:
or would overwriting `false` make a mess when we have:

* `max-age:10`
* `s-maxage:20`

currently the result would be `max-age:10` 

but if we overwrite the `false` for s-maxage, we would end up with `max-age: 10 s-maxage:20` which is more lax. your change would change

* `public max-age:10`
* `max-age: 30 s-maxage:20`

to be `max-age:10 s-maxage:20` (public is dropped because the second fragment is not marked public)

Reply:
Not sure I understand the question… I think the code should work exactly as you describe.

Reply:
> if one fragment does not specify max-age, but others do, don't we want the lowest value of max-age output, rather than no max-age at all?

That's the goal of this PR, as long as the fragment without max-age is `public`.

Reply:
> your change would change
> `public max-age:10`
> `max-age: 30 s-maxage:20`
> to be `max-age:10 s-maxage:20` (public is dropped because the second fragment is not marked public)

No, it gives `public, max-age=10` because 

1. `public` is implied by `s-maxage` in the second response and thus everything is `public`,
2. the minimum of all provided lifetimes applying for a shared cache is `10`,
3. `s-maxage` and `max-age` are reduced to `max-age` only when the values are the same.

@dbu If we agree this is correct and if you think it's worthwile a dedicated test, I will add one.

Reply:
sorry, i missed the implied public thing. `public max-age:10` means that the shared cache may also use a max age of 10, right? then that seems correct, and i think its a good idea to have a test for this to explain how it is supposed to work.

Reply:
@dbu Test added.